### ISSUE TITLE
ifx 2024.0.2 and earlier compiler bug affecting SWALS use of NULL procedure pointer

### ISSUE THREAD

**Issue by gareth-d-ga** [GeoscienceAustralia/ptha#40]
The `ifort` compiler will be discontinued in 2024, and we need to move to compiling SWALS using `ifx` (or an alternative such as `gfortran` that may be less optimized on NCI).

For awhile now `ifx` has a bug in SWALS that occurs if we use a `forcing_patch_type`. It tends to segfault [around here](https://github.com/GeoscienceAustralia/ptha/blob/52502a775377cc3f776a9756c07c916949643cdd/propagation/SWALS/src/shallow_water/forcing_mod.f90#L120).
* I confirmed this still exists in ifx `2024.0.2`.
* In typical usage the `forcing_time_function => NULL()`, matching how it is initialised in the derived type [here](https://github.com/GeoscienceAustralia/ptha/blob/52502a775377cc3f776a9756c07c916949643cdd/propagation/SWALS/src/shallow_water/forcing_mod.f90#L36). 
  * The code uses `associated(forcing_time_function)` to determine whether to call the function, or use a default calculation. 
  * The compiler seems to have a problem with the call to `associated(forcing_time_function)`.  
  * If we comment out this line and other nearby code in the if statement, leaving just the default treatment, then the segfault dissappears. 

The above comment shows a workaround for most current SWALS use cases. 

But the code works with `gfortran` and `ifort`, and has done so for a long time. So it would be good to get this issue fixed.

**Comment by gareth-d-ga** [comment]
You could also try a workaround, e.g. making the procedure pointer an optional argument, then checking the status in a parent routine, and branching at that point. This seems less elegant, but worth trying.

**Comment by gareth-d-ga** [comment]
To be clear, here's the workaround that can be used to make the tests pass with `ifx`. Notice the tests are currently only exercising the "default branch" of this if statement, where the procedure pointer is not associated (`NULL()`). The "non-default" branch has also been used successfully with ifort/gfortran in some experimental models, but I haven't tried it with ifx.

```
diff --git a/propagation/SWALS/src/shallow_water/forcing_mod.f90 b/propagation/SWALS/src/shallow_water/forcing_mod.f90
index 05af7ccb..156e4fe3 100644
--- a/propagation/SWALS/src/shallow_water/forcing_mod.f90
+++ b/propagation/SWALS/src/shallow_water/forcing_mod.f90
@@ -117,21 +117,21 @@ module forcing_mod
             else
                 ! The time-step might not be completely between start_time and end_time.
                 ! Determine the fraction of the forcing we actually need to apply
-                if(.not. associated(forcing_time_function)) then
+                !@if(.not. associated(forcing_time_function)) then
                     ! Default case: Constant forcing
                     update_fraction = (min(time, end_time) - max((time-dt), start_time))/(end_time - start_time)
-                else
-                    ! User-specified function
-                    p2 = ( min(time, end_time)        - start_time ) / (end_time - start_time)
-                    p1 = ( max(time - dt, start_time) - start_time ) / (end_time - start_time)
-                    update_fraction = forcing_time_function(p2) - forcing_time_function(p1)
-                    if(update_fraction < 0.0_dp .or. update_fraction > 1.0_dp) then
-                        write(log_output_unit, *) &
-                            'Error: forcing_time_function must be non-decreasing and map [0.0, 1.0] -> [0.0,1.0]', &
-                            p2, p1, forcing_time_function(p2), forcing_time_function(p1), update_fraction
-                        call generic_stop
-                    end if
-                end if
+                !@else
+                !@    ! User-specified function
+                !@    p2 = ( min(time, end_time)        - start_time ) / (end_time - start_time)
+                !@    p1 = ( max(time - dt, start_time) - start_time ) / (end_time - start_time)
+                !@    update_fraction = forcing_time_function(p2) - forcing_time_function(p1)
+                !@    if(update_fraction < 0.0_dp .or. update_fraction > 1.0_dp) then
+                !@        write(log_output_unit, *) &
+                !@            'Error: forcing_time_function must be non-decreasing and map [0.0, 1.0] -> [0.0,1.0]', &
+                !@            p2, p1, forcing_time_function(p2), forcing_time_function(p1), update_fraction
+                !@        call generic_stop
+                !@    end if
+                !@end if
             end if
             !print*, update_fraction
```

**Comment by gareth-nx** [comment]
Notes below from a colleague who analysed the binary that caused the crash

> Yeah, I think there's a compiler bug there. I disassembled the binary and I think it thinks there's an extra level of indirection. I.e. it's treating the value as a pointer to a pointer to a function:
```
  4cda3d:       48 8b 4d 80             mov    rcx,QWORD PTR [rbp-0x80]
  4cda41:       48 8b 09                mov    rcx,QWORD PTR [rcx]
  4cda44:       62 f1 66 08 c2 ca 02    vcmpless k1,xmm3,xmm2
  4cda4b:       62 f1 5e 08 c2 d1 02    vcmpless k2,xmm4,xmm1
  4cda52:       c5 f8 28 c4             vmovaps xmm0,xmm4
  4cda56:       62 f1 7e 0a 10 c1       vmovss xmm0{k2},xmm0,xmm1
  4cda5c:       c5 ea 5c cc             vsubss xmm1,xmm2,xmm4
  4cda60:       62 f1 6e 09 10 d3       vmovss xmm2{k1},xmm2,xmm3
  4cda66:       48 85 c9                test   rcx,rcx
```
> The first two lines are load an address from memory and then load the value at that address. Then the floating-point instructions are some bits that are common to both branches of that IF statement (so it's pulled them up to before the actual test), and then the final line is the actual test for if the pointer is associated or not (i.e. not 0). The second line is where it's segfaulting, which is why I think it's thinking there's an extra level of indirection (i.e. the first line actually loaded the address of the function rather than the address of a pointer to the the function – since that pointer would be 0 as it's not associated, the second line segfaulted).

> That code is assuming that function pointer was pass-by-reference (normal for Fortran), however earlier in the code I can see it copying the value of the function pointer to that memory location rather than its address. So I think it may have got confused when it inlined add_forcing_work_to_U_over_time. At least that's what I think has happened: as is common for Fortran, the compiled machine code is actually quite verbose and convoluted.

**Comment by gareth-d-ga** [comment]
I created a [pull request](https://github.com/GeoscienceAustralia/ptha/pull/42) to workaround the issue. This was done by making procedure pointer an optional argument, then checking the status in a parent routine, and branching at that point. 

This leads to a few more lines of code, but that's probably the easiest solution for now.

**Comment by gareth-d-ga** [comment]
Fixed by #42
